core/state/snapshot/iterator_binary.go

```diff
@@ -117,33 +90,69 @@ func (dl *diffLayer) initBinaryStorageIterator(account common.Hash) Iterator {
 // or an error if iteration failed for some reason (e.g. root being iterated
 // becomes stale and garbage collected).
 func (it *binaryIterator) Next() bool {
+	// verify ensures that the item at the iterator current position exists.
+	// The `it.fail` will be set if any error occurs, check the error later.
+	verify := func() bool {
+		if it.accountIterator {
+			return len(it.Account()) != 0
+		}
```

Comment:
Isn't it a bit odd, though, that the underlying account/storage iterator returns a non-existing element?

Reply:
the underlying account/storage iterator returns the items, which represent the deleted element (deleted storage slot or deleted account)

Reply:
So, 

- If it is expected that an AccountIterator / StorageIterator should yield deleted values, then this filtering done in the `binaryIterator` breaks that assumption, and thus does not implement a proper iterator. 
- If it is not expected that account/storage iterators yield deleted items, then the underlying iterators that are used here are violating the contract. 

Not sure how to consider this case. You could just filter on the outside: make a `LiveIterator` which calls `next` on the underlying iterator as long as it keeps returning delete stuff. That might be the cleanest, IMO

Reply:
How does the `fastIterator` handle deletes? Does it emit them?

Reply:
Ah, from a comment inside the fast iterator
```
	// If an account or a slot is deleted in one of the layers, the key will
	// still be there, but the actual value will be nil. However, the iterator
	// should not export nil-values (but instead simply omit the key), so we
	// need to loop here until we either
	//  - get a non-nil value,
	//  - hit an error,
	//  - or exhaust the iterator
```

So the change you're doing here seems ok. This is a lack of test coverage though!

Reply:
Yep, fastIterator has similar logic to skip the deleted items internally.